RFC-0001-prsestissimo-metrics.md

```diff
@@ -0,0 +1,251 @@
+# Prestissimo Runtime Metrics Capturing and reporting
+
+## Proposers
+
+* Ahana, an IBM company.
+* Uber.
+  
+## Prototype PR
+There are two approaches in this PR:https://github.com/prestodb/presto/pull/21599 which will be discussed in detail.
+
+## Summary
+
+Propose design to capture Prestissimo runtime metrics and expose these metrics to a popular time series DB, Prometheus.
+Also, discuss a general way to capture and serialize metrics to any custom format.
+
+## Background
+### Velox runtime metrics:
+Velox (the C++ runtime framework) defines a set of runtime metrics to record events that give insights into worker health
+status like CPU, memory, cache hits etc. Velox exposes the BaseStatsReporter interface to capture these runtime metrics.
+It is applications’ responsibility to capture these runtime events for monitoring purposes. The Velox system defines the
+following types of metrics: Count, Sum, Avg, Rate and Histogram. Refer this document: [Velox Runtime metrics](https://github.com/facebookincubator/velox/blob/main/velox/docs/metrics.rst) for a
+complete list of supported Velox runtime metrics and their definitions. There are a total of **28 metrics** as of Jan 2024
+declared [here](https://github.com/facebookincubator/velox/blob/main/velox/common/base/Counters.h) in the Velox code.
+
+### PrestoCPP Counters:
+In addition to Velox metrics, Prestissimo (the C++ Presto Worker) also defines worker related Prestissimo Counters [here](https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/main/common/Counters.h). There are **102 counters**
+defined as of Jan 2024. These metrics are periodically reported using the BaseStatsReporter interface as well.
+Additionally, Velox exposes the following interfaces to capture task or expression completion events:
+
+### Other Mechanisms
+***TaskListener:*** This interface exposes `TaskListener::onTaskCompletion()` method that is called on completion of each task
+on a specific worker. The task specific metrics are passed down to this method.
+
+***ExprSetListener:*** Similar to TaskListener, this interface exposes `ExprSetListener::onCompletion()` interface that must
+be implemented by the user in order to capture Expression level metrics.
+
+The above mechanisms provide extremely granular stats that provide more context about a query, task or expression behaviors.
+
+### Prometheus Time Series DB:
+Open source Time Series DB(TDB) for capturing and storing time series events. Prometheus supports COUNT,
```

Comment:
Its worth noting that a lot of other timeseries dbs accept metrics in prometheus format. Hence this section can maybe be refactored to talk a bit about the format. 
Also some of the timeseries dbs which accept metrics in prometheus format accept a push model. The library can be used for that as well. We need to define if we will enable a push model using the library through some configurations or state that this can be supported in future / custom cases

Reply:
Done.